src/GafferScene/InteractiveRender.cpp

```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+
+class InteractiveRender::SceneGraphIteratorFilter : public tbb::filter
+{
+	public:
+		SceneGraphIteratorFilter( InteractiveRender::SceneGraph *start ) :
+			tbb::filter( tbb::filter::serial_in_order ), m_current( start )
+		{
+			m_childIndices.push_back( 0 );
+		}
+		
+		void next()
+		{
+			// go down one level in the hierarchy if we can:
+			if( m_current->m_children.size() )
+			{
+				m_current = m_current->m_children[0];
+				m_childIndices.push_back(0);
+				return;
+			}
+			
+			while( m_childIndices.size() )
+			{
+				
+				// increment child index:
+				++m_childIndices.back();
+				
+				// find parent's child count - for the root we define this as 1:
+				size_t parentNumChildren = m_current->m_parent ? m_current->m_parent->m_children.size() : 1;
+				
+				if( m_childIndices.back() == parentNumChildren )
+				{
+					// we've got to the end of the child list, jump up one level:
+					m_childIndices.pop_back();
+					m_current = m_current->m_parent;
+					continue;
+				}
+				else
+				{
+					// move to next child of the parent:
+					m_current = m_current->m_parent->m_children[ m_childIndices.back() ];
+					return;
+				}
+			}
+		}
+		
+		void *operator()( void *item )
+		{
+			if( m_childIndices.empty() )
+			{
+				// we've finished the iteration
+				return NULL;
+			}
+			InteractiveRender::SceneGraph *s = m_current;
+			next();
+			return s;
+		}
+	
+	private:
+		
+		SceneGraph *m_current;
+		std::vector<size_t> m_childIndices;
+};
+
+
+//////////////////////////////////////////////////////////////////////////
+// SceneGraphEvaluatorFilter implementation
+//
+// This parallel filter computes the data living at the scene graph
+// location it receives. If the "onlyChanged" flag is set to true, it
+// only recomputes data when the hashes change, otherwise it computes
+// all non null scene data.
+//
+//////////////////////////////////////////////////////////////////////////
+
+class InteractiveRender::SceneGraphEvaluatorFilter : public tbb::filter
+{
+	public:
+		SceneGraphEvaluatorFilter( const ScenePlug *scene, const Context *context, bool onlyChanged ) :
+			tbb::filter( tbb::filter::parallel ), m_scene( scene ), m_context( context ), m_onlyChanged( onlyChanged )
+		{
+		}
+
+		void *operator()( void *item )
+		{
+			SceneGraph *s = (SceneGraph*)item;
+			ScenePlug::ScenePath path;
+			s->path( path );
+
+			try
+			{
+				ContextPtr context = new Context( *m_context, Context::Borrowed );
+				context->set( ScenePlug::scenePathContextName, path );
+				Context::Scope scopedContext( context.get() );
+			
+				if( m_onlyChanged )
+				{
+					// we're re-traversing this location, so lets only recompute attributes where
+					// their hashes change:
+
+					IECore::MurmurHash attributesHash = m_scene->attributesPlug()->hash();
+					if( attributesHash != s->m_attributesHash )
+					{
+						s->m_attributesHash = attributesHash;
+						s->m_attributes = m_scene->attributesPlug()->getValue();
```

Comment:
The `getValue()` method accepts an optional precomputed hash, which can significantly speed things up (because otherwise `getValue()` will compute the hash again itself). We should definitely use that.